balancer/rls/internal/client.go

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package rls
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	rlspb "google.golang.org/grpc/balancer/rls/internal/proto/grpc_lookup_v1"
+)
+
+// For gRPC services using RLS, the value of target_type in the
+// RouteLookupServiceRequest will be set to this.
+const grpcTargetType = "grpc"
+
+// rlsClient is a simple wrapper around a RouteLookupService client which
+// provides non-blocking semantics on top of a blocking unary RPC call.
+//
+// The RLS LB policy creates a new rlsClient object with the following values:
+// * a grpc.ClientConn to the RLS server using appropriate credentials from the
+//   parent channel
+// * dialTarget corresponding to the original user dial target, e.g.
+//   "firestore.googleapis.com".
+//
+// The RLS LB policy uses an adaptive throttler to perform client side
+// throttling and asks this client to make an RPC call only after checking with
+// the throttler.
+type rlsClient struct {
+	cc         *grpc.ClientConn
+	stub       rlspb.RouteLookupServiceClient
+	dialTarget string
+	rpcTimeout time.Duration
```

Comment:
The reason I decided to just take the `rpcTimeout` here was so that I could build the context with this timeout value and return the cancel function to the caller so that they can cancel early if they want to.

If I follow your suggestion, the the `lookup` API will take a `context` object, but will not return anything. The caller can just decide to cancel the context whenever they want. Do you see any outright reasons why one might be better than the other?

Reply:
So, now that the code dealing with pending request cancellation is removed, I think it is easier for the caller if they dont have to allocate a context object and clean it up everytime they make an RLS call. It would be better handled here I guess. Wdyt?

Reply:
Is the timeout configured by the balancer config? Or it will always be the default value?
Can we remove it from the struct (we probably still need to do something for testing though.)

Reply:
It is configured by the balancer config.

And since this client object is really a thin wrapper, my idea is that everytime the lbConfig changes (even when just the timeout value changes, the RLS server name remains the same), the LB policy can create a new client object.

Reply:
Also, given how simple the client is now, and given that the client and the lb policy will be in the same package, this client is more about making the code more easier to read and test. Do I make sense?

Reply:
Although, there will be one more piece of functionality added to this client, which will justify having this client object instead of just spawning off a goroutine from the lb policy itself and handling the response via a callback.

That piece of functionality will monitor the connectivity state of this client. Again, nothing prevents us from doing all that in the lb policy code as well. Although splitting it out seems nicer :)